src/llfs/ioring_log_device2.hpp

```diff
@@ -0,0 +1,498 @@
+
+  /** \brief I/O callback that handles the completion of a write to the control block.
+   */
+  void handle_control_block_update(StatusOr<i32> result) noexcept;
+
+  /** \brief Handles all write errors.
+   */
+  void handle_write_error(Status status) noexcept;
+
+  /** \brief Returns a wrapped handler for async writes.
+   *
+   * Automatically injects `this` as the first arg to `handler`, so that callers need not capture
+   * this in handler itself.
+   */
+  template <typename HandlerFn>
+  auto make_write_handler(HandlerFn&& handler);
+
+  /** \brief Allocates an array of HandlerMemoryStorage objects, adding each to the free pool linked
+   * list (this->handler_memory_pool_).
+   */
+  void initialize_handler_memory_pool();
+
+  /** \brief Pops the next HandlerMemoryStorage object off this->handler_memory_pool_ and uses it to
+   * construct a new HandlerMemory object, returning a pointer to the newly constructed
+   * HandlerMemory.
+   *
+   * This function MUST only be called on the event loop thread.
+   */
+  HandlerMemory* alloc_handler_memory() noexcept;
+
+  /** \brief Destructs the passed HandlerMemory object and adds its storage back to the pool.
+   *
+   * This function MUST only be called on the event loop thread.
+   */
+  void free_handler_memory(HandlerMemory* p_mem) noexcept;
+
+  //+++++++++++-+-+--+----- --- -- -  -  -   -
+
+  /** \brief The context passed in at construction time; provides access to the ring buffer.
+   */
+  LogStorageDriverContext& context_;
+
+  /** \brief The configuration of the log; passed in at construction time.
+   */
+  const IoRingLogConfig2 config_;
+
+  /** \brief Runtime configuration options, such as the maximum number of concurrent writes allowed.
+   */
+  const LogDeviceRuntimeOptions options_;
+
+  /** \brief Provides access to the underlying storage media.
+   */
+  StorageT storage_;
+
+  /** \brief The minimum IO page size for the storage media.
+   */
+  const usize device_page_size_ = usize{1} << this->config_.device_page_size_log2;
+
+  /** \brief The size to which data writes must be aligned; this is also the distance from the start
+   * of the control block to the start of the log data storage area.  Typically aligns with memory
+   * pages (4k by default).
+   */
+  const usize data_page_size_ = usize{1} << this->config_.data_alignment_log2;
+
+  /** \brief The absolute file (media) offset of the beginning of the data region.
+   */
+  i64 data_begin_;
+
+  /** \brief The absolute file (media) offset of the end (non-inclusive) of the data region.
+   */
+  i64 data_end_;
+
+  /** \brief True iff storage_.on_work_started() has been called (inside open).
+   */
+  std::atomic<bool> work_started_{false};
+
+  /** \brief True iff halt has been called.
+   */
+  std::atomic<bool> halt_requested_{false};
+
+  /** \brief Stores the two externally modifiable Watch objects (target trim pos and commit pos).
```

Comment:
They are "strong typedefs" (https://gitlab.com/batteriescpp/batteries/-/blob/main/src/batteries/strong_typedef.hpp?ref_type=heads#L85).  In this case, the types are used as indices into a container that stores multiple `batt::Watch` objects, one for each pointer.  This allows certain methods like `observe` and `wait_for_slot_offset_change` to have generic (templated) implementations, instead of having the same code pattern copy-pasted.